pkg/image/strategy/simplelookup.go

```diff
@@ -0,0 +1,221 @@
+		// ie, if it is coming from anywhere internal, it should have an ID, if not, then it's
+		// coming straight from user.
+		if len(ref.ID) != 0 {
+			dgst, err := digest.Parse(ref.ID)
+			if err != nil {
+				continue
+			}
+			_, err = manifests.Get(ctx, dgst)
+			if err != nil {
+				continue
+			}
+		}
+		altRef.ID = ref.ID
+		altRef.Tag = ref.Tag
+		return altRepo, altRef, nil
+	}
+	return nil, ref, fmt.Errorf("did not find alternative image reference for %s", ref.String())
+}
+
+// addICSPsFromCluster will lookup ImageContentSourcePolicy resources in cluster.
+func (s *simpleLookupICSP) addICSPsFromCluster(ctx context.Context) error {
+	if s.icspClient == nil {
+		return fmt.Errorf("no client to access ImageContentSourcePolicies in cluster")
+	}
+	icsps, err := s.icspClient.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		// may or may not have access to ICSPs in cluster
+		// don't error if can't access ICSPs
+		return fmt.Errorf("did not access any ImageContentSourcePolicies in cluster: %v", err)
+	}
+	if len(icsps.Items) == 0 {
+		return fmt.Errorf("no ImageContentSourcePolicies found in cluster")
+	}
+	s.icspList = append(s.icspList, icsps.Items...)
+	return nil
+}
+
+// addImageSourcePoliciesFromFile appends to list of alternative image sources from ICSP file
+// returns error if no icsp object decoded from file data
+func (s *simpleLookupICSP) addImageSourcePoliciesFromFile() error {
+	if len(s.icspFile) == 0 {
+		return nil
+	}
+	icspData, err := ioutil.ReadFile(s.icspFile)
+	if err != nil {
+		return fmt.Errorf("unable to read ImageContentSourceFile %s: %v", s.icspFile, err)
+	}
+	if len(icspData) == 0 {
+		return fmt.Errorf("no data found in ImageContentSourceFile %s", s.icspFile)
+	}
+	icspObj, err := runtime.Decode(operatorv1alpha1scheme.Codecs.UniversalDeserializer(), icspData)
+	if err != nil {
+		return fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", s.icspFile, err)
+	}
+	icsp, ok := icspObj.(*operatorv1alpha1.ImageContentSourcePolicy)
+	if !ok {
+		return fmt.Errorf("could not decode ImageContentSourcePolicy from %s", s.icspFile)
+	}
+	s.icspList = append(s.icspList, *icsp)
+	return nil
+}
+
+// alternativeImageSources returns unique list of DockerImageReference objects from list of ImageContentSourcePolicy objects
+func (s *simpleLookupICSP) alternativeImageSources(imageRef reference.DockerImageReference) ([]reference.DockerImageReference, error) {
+	var imageSources []reference.DockerImageReference
+	for _, icsp := range s.icspList {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		var sourceMatches bool
+		for _, rdm := range repoDigestMirrors {
+			rdmRef, err := reference.Parse(rdm.Source)
+			if err != nil {
+				return nil, err
+			}
+			if imageRef.AsRepository() == rdmRef.AsRepository() {
+				klog.V(2).Infof("%v RepositoryDigestMirrors source matches given image", imageRef.AsRepository())
+				sourceMatches = true
```

Comment:
nit: this could be easier to read and a little bit faster if we inverted the logic to something:

```golang
if imageRef.AsRepository() != rdmRef.AsRepository() {
    continue
}

klog.V(2).Infof("%v RepositoryDigestMirrors source matches given image", imageRef.AsRepository())
```

We would avoid a unnecessary run over the `Mirrors` slice and the `sourceMatches` would not be needed anymore.

Reply:
In fact seems like I might have hit an issue with the logic in here, taking into account the following ImageContentSourcePolicy :

```yaml
apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  name: goiaba
spec:
  repositoryDigestMirrors:
  - source: quay.io/rmarasch/does-not-exist
    mirrors:
    - goiaba.news:5000/rmarasch/simple-web-server
  - source: docker.io/rmarasch/does-not-exist
    mirrors:
    - i.do.not.exist/rmarasch/simple-web-server
```

When pulling `quay.io/rmarasch/does-not-exist` I would expect to have as an alternative `goiaba.news:5000/rmarasch/simple-web-server` but I am seeing `i.do.not.exist/rmarasch/simple-web-server` instead.

Reply:
I have a unit test for that: [test icspfile](https://github.com/openshift/library-go/blob/076e9be4cebf1a7f9b8c8434967dd058bac8622e/pkg/image/strategy/simplelookup_test.go#L14-#L39)  and [test case](https://github.com/openshift/library-go/blob/076e9be4cebf1a7f9b8c8434967dd058bac8622e/pkg/image/strategy/simplelookup_test.go#L96-#L97)  but hm, if you are seeing different I'll take a closer look.

Reply:
I can't reproduce this issue anymore, please disregard this comment.